contracts/ibc-reflect-send/src/contract.rs

```diff
@@ -0,0 +1,649 @@
+    deps: DepsMut,
+    info: MessageInfo,
+    new_admin: HumanAddr,
+) -> StdResult<HandleResponse> {
+    // auth check
+    let mut cfg = config(deps.storage).load()?;
+    if info.sender != cfg.admin {
+        return Err(StdError::generic_err("Only admin may set new admin"));
+    }
+    cfg.admin = new_admin;
+    config(deps.storage).save(&cfg)?;
+
+    Ok(HandleResponse {
+        messages: vec![],
+        attributes: vec![
+            attr("action", "handle_update_admin"),
+            attr("new_admin", cfg.admin),
+        ],
+        data: None,
+    })
+}
+
+pub fn handle_send_msgs(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    channel_id: String,
+    msgs: Vec<CosmosMsg>,
+) -> StdResult<HandleResponse> {
+    // auth check
+    let cfg = config(deps.storage).load()?;
+    if info.sender != cfg.admin {
+        return Err(StdError::generic_err("Only admin may send messages"));
+    }
+    // ensure the channel exists (not found if not registered)
+    accounts(deps.storage).load(channel_id.as_bytes())?;
+
+    // construct a packet to send
+    let packet = PacketMsg::Dispatch { msgs };
+    let msg = IbcMsg::SendPacket {
+        channel_id,
+        data: to_binary(&packet)?,
+        timeout_block: None,
+        timeout_timestamp: build_timeout(&env.block),
+    };
+
+    Ok(HandleResponse {
+        messages: vec![msg.into()],
+        attributes: vec![attr("action", "handle_send_msgs")],
+        data: None,
+    })
+}
+
+pub fn handle_check_remote_balance(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    channel_id: String,
+) -> StdResult<HandleResponse> {
+    // auth check
+    let cfg = config(deps.storage).load()?;
+    if info.sender != cfg.admin {
+        return Err(StdError::generic_err("Only admin may send messages"));
+    }
+    // ensure the channel exists (not found if not registered)
+    accounts(deps.storage).load(channel_id.as_bytes())?;
+
+    // construct a packet to send
+    let packet = PacketMsg::Balances {};
+    let msg = IbcMsg::SendPacket {
+        channel_id,
+        data: to_binary(&packet)?,
+        timeout_block: None,
+        timeout_timestamp: build_timeout(&env.block),
+    };
+
+    Ok(HandleResponse {
+        messages: vec![msg.into()],
+        attributes: vec![attr("action", "handle_check_remote_balance")],
+        data: None,
+    })
+}
+
+pub fn handle_send_funds(
+    deps: DepsMut,
+    env: Env,
+    mut info: MessageInfo,
+    reflect_channel_id: String,
+    transfer_channel_id: String,
+) -> StdResult<HandleResponse> {
+    // intentionally no auth check
+
+    // require some funds
+    if info.sent_funds.is_empty() {
+        return Err(StdError::generic_err(
+            "you must send the coins you with to ibc transfer",
+        ));
+    } else if info.sent_funds.len() > 1 {
+        return Err(StdError::generic_err("you can only ibc transfer one coin"));
+    }
+    let amount = info.sent_funds.swap_remove(0);
```

Comment:
I'm curious, how is this better or different than `info.sent_funds[0]`?

Reply:
> different than info.sent_funds[0]?

You get an owned `Coin`, which avoid a heap allocation on `Coin::denom` when cloning it.

However, you get the same by using pop, which is simpler to review in my opinion:
1. Check `info.sent_funds.len() > 1` and error as above
2. `let amount = match info.sent_funds.pop() { …` which errors out if there is nothing to pop